app/1.0.x/plugin-development/custom-entities.md

```diff
@@ -129,76 +129,400 @@ To see a real-life example, give a look at the [Key-Auth plugin migrations](http
-relates to in the datastore, constraints on its fields such as foreign keys,
-non-null constraints and such.
+The first step to using custom entities in a custom plugin is defining one
+or more *schemas*.
 
-This schema is to be defined in a module named:
+A schema is a Lua table which describes entities. There's structural information
+like how are the different fields of the entity named and what are their types,
+which is similar to the fields describing your [plugin
+configuration]({{page.book.chapters.plugin-configuration}})).
+Compared to plugin configuration schemas, custom entity schemas require
+additional metadata (e.g. which field, or fields, constitute the entities'
+primary key).
+
+Schemas are to be defined in a module named:
 
 ```
-"kong.plugins.<plugin_name>.daos"
+kong.plugins.<plugin_name>.daos
 ```
 
-Once that module returns your entity's schema, and assuming your plugin is
-loaded by Kong (see the `plugins` property in `kong.conf`), the DAO Factory
-will use it to instantiate a DAO object.
-
-Here is an example of how one would define a schema to store API keys in a his
-or her database:
+Meaning that there should be a file called `<plugin_name>/daos.lua` inside your
+plugin folder. The `daos.lua` file should return a table containing one or more
+schemas. For example:
 
 ```lua
--- daos.lua
-local SCHEMA = {
-  primary_key = {"id"},
-  table = "keyauth_credentials", -- the actual table in the database
+-- <plugin_name>/daos.lua
+local typedefs = require "kong.db.schema.typedefs"
+
+local credentials = {
+  primary_key = { "id" },
+  name = "keyauth_credentials",
+  endpoint_key = "key",
+  cache_key = { "key" },
   fields = {
-    id = {type = "id", dao_insert_value = true}, -- a value to be inserted by the DAO itself (think of serial ID and the uniqueness of such required here)
-    created_at = {type = "timestamp", immutable = true, dao_insert_value = true}, -- also interted by the DAO itself
-    consumer_id = {type = "id", required = true, foreign = "consumers:id"}, -- a foreign key to a Consumer's id
-    key = {type = "string", required = false, unique = true} -- a unique API key
+    { id = typedefs.uuid },
+    { created_at = typedefs.auto_timestamp_s },
+    { consumer = { type = "foreign", reference = "consumers", default = ngx.null, on_delete = "cascade" } },
+    { key = { type = "string", required = false, unique = true, auto = true } },
   }
 }
 
-return {keyauth_credentials = SCHEMA} -- this plugin only results in one custom DAO, named `keyauth_credentials`
+return { credentials }
 ```
 
-Since your plugin might have to deal with multiple custom DAOs (in the case
-when you want to store several entities), this module is bound to return a
-key/value table where keys are the name on which the custom DAO will be
-available in the DAO Factory.
+This example `daos.lua` file introduces a single schema called `keyauth_credentials`.
+
+Here is a description of some top-level properties:
+
+<table>
```

Comment:
:O When I said "table", I had "mardown notation table" in mind. Did it take you a long time to write it with HTML? Or is it because it accomplishes something a markdown table couldn't?

Reply:
Markdown tables are, as far as I know, limited to 1 line per row. You can not use things like bullet points or several paragraphs on them, which I wanted to use. Unfortunately once you go the html-table route you lose all the markdown amenities like using apostrophes for `<code>` sections.

It did take some time to type, yes. If I end up doing this sort of thing often in the future, I will have to invest some time optimizing my vim setup for it.